iceoryx_hoofs/test/moduletests/test_containers_uninitialized_array.cpp

```diff
@@ -0,0 +1,228 @@
+
+TYPED_TEST(UninitializedArrayTest, accessElements)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f8621711-9c4c-45b5-979d-404357b664a4");
+    auto& buffer = this->buffer;
+    auto capacity = buffer.capacity();
+
+    int startValue = 73;
+    this->fillBuffer(startValue);
+
+    int expected = startValue;
+    for (uint64_t i = 0; i < capacity; ++i)
+    {
+        EXPECT_EQ(buffer[i], expected++);
+    }
+}
+
+TYPED_TEST(UninitializedArrayTest, accessElementsOfConstUinitializedArray)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "50575bac-cff9-4896-89da-b03753370b18");
+    auto& buffer = this->buffer;
+    auto capacity = buffer.capacity();
+
+    int startValue = 37;
+    this->fillBuffer(startValue);
+
+    const auto& constBuffer = this->buffer;
+    int expected = startValue;
+    for (uint64_t i = 0; i < capacity; ++i)
+    {
+        EXPECT_EQ(constBuffer[i], expected++);
+    }
+}
+
+TEST(UninitializedArray, FirstElementIsInitializedWithZeroWhenBufferSetToFirstElementZeroed)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "bb213516-ab37-43e3-b2ec-098c98d777d1");
+    UninitializedArray<uint32_t, 2, iox::containers::FirstElementZeroed> buffer;
+    EXPECT_THAT(buffer[0], Eq(0));
+}
+
+TYPED_TEST(UninitializedArrayTest, BeginReturnsIteratorToBeginningOfUninitializedArray)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "6434e308-e24f-41e1-a1e1-949da01b2cbb");
+    auto& buffer = this->buffer;
+    EXPECT_THAT(buffer.begin(), Eq(&buffer[0]));
+}
+
+TYPED_TEST(UninitializedArrayTest, ConstBeginReturnsIteratorToBeginningOfUninitializedArray)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "7387b043-db44-47ac-a2da-c40040bb9baa");
+    auto& buffer = this->buffer;
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast) reuse of buffer to test const method
+    EXPECT_THAT(const_cast<const decltype(buffer)>(buffer).begin(), Eq(&buffer[0]));
+}
+
+TYPED_TEST(UninitializedArrayTest, EndReturnsIteratorToEndOfUninitializedArray)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "52447fba-0c7f-40df-8b7f-64d8b3ffcc49");
+    auto& buffer = this->buffer;
+    EXPECT_THAT(buffer.end(), Eq(&buffer[buffer.capacity()]));
+}
+
+TYPED_TEST(UninitializedArrayTest, ConstEndReturnsIteratorToEndOfUninitializedArray)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "2946ad83-b782-4c54-966b-c94b482335cc");
+    auto& buffer = this->buffer;
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast) reuse of buffer to test const method
+    EXPECT_THAT(const_cast<const decltype(buffer)>(buffer).end(), Eq(&buffer[buffer.capacity()]));
+}
+
+TYPED_TEST(UninitializedArrayTest, BeginIteratorComesBeforeEndIteratorWhenNotEmpty)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1a180d15-7e77-4234-ad88-04673cbf9fc9");
+    auto& buffer = this->buffer;
+    buffer[0] = 1;
+    EXPECT_THAT(buffer.begin(), Lt(buffer.end()));
+}
+
+TYPED_TEST(UninitializedArrayTest, BeginConstIteratorComesBeforeEndConstIteratorWhenNotEmpty)
```

Comment:
This maybe true for this particular implementation, but in general STL `end` is only required to be different from all other valid iterators, in particular `begin`. This can be seen by the equivalent iterator version of range-based loops, which check for inequality with.

TLDR: I would suggest we do not impose an order and test this unnecessary restriction, it would suffice to just test inequality of `begin` and `end` for a non-empty array (+ const variations, of course, but const iterators only need to be comparable to other const iterator). Furthermore `begin` should point to `buffer[0]` of course if the array is not-empty and to `end` if it is empty.